BUG: pd.cut with bins=1 and input all 0s

luca-s opened the pull request:
The special case of running pd.cut() qith bins=1 an input containing all
0s raises a ValueError

 - [x] closes #15428
closes #15431 
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master | flake8 --diff``
 - [X] whatsnew entry

jreback commented:
pls add #15428 as a tests case as well

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -587,3 +587,5 @@ Bug Fixes
 - Bug in ``Series.replace`` and ``DataFrame.replace`` which failed on empty replacement dicts (:issue:`15289`)
 - Bug in ``pd.melt()`` where passing a tuple value for ``value_vars`` caused a ``TypeError`` (:issue:`15348`)
 - Bug in ``.eval()`` which caused multiline evals to fail with local variables not on the first line (:issue:`15342`)
+
```
don't put whatsnew at the very end as they cause conflicts. the point of the empty space elsewhere is for you to use

luca-s replied:
Thanks, I was actually wondering what was the best place to add the new entry

jreback commented on doc/source/whatsnew/v0.20.0.txt:
```diff
@@ -587,3 +587,5 @@ Bug Fixes
 - Bug in ``pd.melt()`` where passing a tuple value for ``value_vars`` caused a ``TypeError`` (:issue:`15348`)
 - Bug in ``.eval()`` which caused multiline evals to fail with local variables not on the first line (:issue:`15342`)
+
+- Bug in ``pd.cut()`` (:issue:`15428`)
```
make this a useful message

jreback commented on pandas/tests/tools/test_tile.py:
```diff
@@ -297,6 +297,10 @@ def test_single_bin(self):
         result = cut(s, 1, labels=False)
         tm.assert_series_equal(result, expected)
 
+        s = Series([0., 0.])
```
add as a new test function
add the issue number as a comment

luca-s replied:
This PR is related to that test. The test checks that pd.cut works fine when the input array contains identical values and bins=1. That is the general scenario, while this PR covers the specific case of an array contains all 0s. it would be misleading to move the test to another function.

luca-s committed:
BUG: pd.cut with bins=1 and input all 0s

The special case of running pd.cut() qith bins=1 an input containing all
0s raises a ValueError

luca-s commented:
Regarding the issue #15428 I can add that fix here too but It's a different code bug and it's related to another function (qcut vs cut).

jreback commented on pandas/tests/tools/test_tile.py:
```diff
@@ -297,6 +297,11 @@ def test_single_bin(self):
         result = cut(s, 1, labels=False)
         tm.assert_series_equal(result, expected)
 
+        # issue 15428
```
add this for also 2 equal values that are not 0, but with a single bin

luca-s replied:
That's already covered

jreback replied:
ok great

jreback commented:
> Regarding the issue #15428 I can add that fix here too but It's a different code bug and it's related to another function (qcut vs cut).

its the same thing. yes please add it.

luca-s commented:
Ok then, I will add that fix too.

luca-s commented:
I am a little confused, it it ok now or is there something still missing?

luca-s commented:
I am referring to #15431, do you want that issue solved in this PR too, even though it's a different bug?

jreback commented:
@luca-s yes its virtually the same issue.

luca-s committed:
Issue #15431

BUG: pd.qcut with q=1 and input with identical values

jreback commented on pandas/tools/tile.py:
```diff
@@ -189,6 +189,13 @@ def qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise'):
         quantiles = q
     bins = algos.quantile(x, quantiles)
+
+    # fix special case: q=1 and all identical values
```
this seems a bit hacky, any way to figure this out a-prior?

luca-s commented:
What about this approach? I added an 'allow' option to to _bins_to_cuts function. This might actually be useful as a qcut option too

luca-s committed:
Added 'allow' duplicates option to _bins_to_cuts

jreback commented on pandas/tools/tile.py:
```diff
@@ -186,9 +186,13 @@ def qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise'):
     if is_integer(q):
         quantiles = np.linspace(0, 1, q + 1)
+
+        if q == 1:
```
this is too narrow of a solution as well

jreback commented on pandas/tools/tile.py:
```diff
@@ -185,6 +185,13 @@ def qcut(x, q, labels=None, retbins=False, precision=3, duplicates='raise'):
+            raise ValueError('Cannot qcut empty array')
+
+        rng = (nanops.nanmin(x), nanops.nanmax(x))
+        if rng[0] == rng[1] and q == 1:
```
instead of using ``allow`` I would just adjust the quantiles right here

luca-s replied:
Sure, I can do that but this seems cleaner than increasing the edges by 0.1% for this specific case only (q=1 and identical values). That seems fine with pd.cut but not with pd.qcut, considering that we might be returning those bins (it's not hidden to the user).

Also, I thought that with a little extra work we could make 'allow' a top level pd.qcut option. Compared to 'drop' this option would maintain the total number of quantiles (some of them empty in the case of duplicated bin edges).

What do you think? If you don't like it I'll go with the 0.1% edges increment.

luca-s replied:
Any update?

jreback replied:
can you show an example of what ``allow`` would do and why its useful?

luca-s replied:
The only reason it could be useful is to keep the number of quantiles as per user request. This is the only difference compare to 'drop'. I understand this might not be super useful, so it's up to you. Still I would keep the option for internal use.

jreback replied:
@luca-s I still find this a bit odd, can you do this by simply adjusting the values (and not using this kw, even though its only internal)?

jreback commented:
you'll want to rebase on master as travis was having some issues

luca-s committed:
Improved solution: using same approach as pd.cut

luca-s committed:
Yet another implementation attempt

luca-s committed:
rebased on master

luca-s commented:
What about this new fix attempt?

jreback commented:
@luca-s I like this a lot more!

luca-s commented:
All good then? Or is there something missing?

jreback commented:
thanks @luca-s 

keep em coming!
